mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp

```diff
@@ -461,6 +462,33 @@ class CastPattern final : public ConvertOpToLLVMPattern<sycl::SYCLCastOp> {
   }
 };
 
+class BarePtrCastPattern final
+    : public ConvertOpToLLVMPattern<sycl::SYCLCastOp> {
+public:
+  using ConvertOpToLLVMPattern<sycl::SYCLCastOp>::ConvertOpToLLVMPattern;
+
+  LogicalResult
+  matchAndRewrite(SYCLCastOp op, OpAdaptor opAdaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    const auto srcType = op.getSource().getType().cast<MemRefType>();
+    const auto resType = op.getResult().getType().cast<MemRefType>();
+    const auto convSrcType = typeConverter->convertType(srcType);
+    const auto convResType = typeConverter->convertType(resType);
+
+    // Ensure the input and result types are legal.
+    if (!canBeLoweredToBarePtr(srcType) || !canBeLoweredToBarePtr(resType) ||
+        !convSrcType || !convResType)
+      return failure();
+
+    Location loc = op.getLoc();
+    LLVMBuilder builder(rewriter, loc);
+    rewriter.replaceOpWithNewOp<LLVM::BitcastOp>(op, convResType,
```

Comment:
Note: what happen if the input memref is in a non-default address space. How do we preserve the address space ?